examples/speaker_diarization/speaker_diarization/utils.py

```diff
@@ -0,0 +1,244 @@
+# Copyright 2021-2023 Kolena Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import re
+from typing import List
+from typing import Set
+from typing import Tuple
+
+import numpy as np
+import pandas as pd
+from pyannote.core import Annotation
+from pyannote.core import Segment
+from scipy import optimize
+from workflow import GroundTruth
+from workflow import Inference
+
+from kolena.workflow.annotation import LabeledTimeSegment
+from kolena.workflow.annotation import TimeSegment
+
+# allowed error (in seconds) when generating identification and missed speech errors.
+ERROR_THRESHOLD = 0.2
+
+
+def build_speaker_index(inf: List[Tuple[str, float, float]]) -> Set[str]:
+    # borrowed from https://github.com/wq2012/SimpleDER/blob/master/simpleder/der.py
+    speaker_set = sorted({element[0] for element in inf})
+    index = {speaker: i for i, speaker in enumerate(speaker_set)}
+    return index
+
+
+def compute_intersection_length(A: List[Tuple[float, float]], B: List[Tuple[float, float]]) -> float:
+    # borrowed from https://github.com/wq2012/SimpleDER/blob/master/simpleder/der.py
+    max_start = max(A[1], B[1])
+    min_end = min(A[2], B[2])
+    return max(0.0, min_end - max_start)
+
+
+def build_cost_matrix(ref, inf) -> dict:
+    # borrowed from https://github.com/wq2012/SimpleDER/blob/master/simpleder/der.py
+    ref_index = build_speaker_index(ref)
+    inf_index = build_speaker_index(inf)
+    cost_matrix = np.zeros((len(ref_index), len(inf_index)))
+    for ref_element in ref:
+        for hyp_element in inf:
+            i = ref_index[ref_element[0]]
+            j = inf_index[hyp_element[0]]
+            cost_matrix[i, j] += compute_intersection_length(
+                ref_element,
+                hyp_element,
+            )
+    return cost_matrix
+
+
+def realign_labels(ref_df: pd.DataFrame, inf_df: pd.DataFrame):
+    """
+    Aligns speaker labels using linear sum optimiztion
+    """
+    ref = [(row.speaker, row.starttime, row.endtime) for i, row in ref_df.iterrows()]
+    inf = [(row.speaker, row.starttime, row.endtime) for i, row in inf_df.iterrows()]
+    cost_matrix = build_cost_matrix(ref, inf)
+    row_index, col_index = optimize.linear_sum_assignment(-cost_matrix)
```

Comment:
This is very cool, @billbai0102.  This finds the maximum intersections of reference and inference segments?

Reply:
Yup, it matches the ref labels with the inf labels such that mappings are unique and intersections are maximized.